### PR TITLE
NIFI-7139 Add record.error.message on failure of a record reader or w…

### PR DESCRIPTION
**Pull request by shawnweeks**
…riter

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

Adds an attribute record.error.message to failing flow files to expose the error message to down stream processors.

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AbstractRecordProcessor.java [hunk 1]
```diff
@@ -174,6 +174,7 @@ public void process(final InputStream in, final OutputStream out) throws IOExcep
             });
         } catch (final Exception e) {
             getLogger().error("Failed to process {}; will route to failure", new Object[] {flowFile, e});
+            session.putAttribute(flowFile,"record.error.message",e.getCause().getLocalizedMessage());
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AbstractRecordProcessor.java]: Is it possible that `getCause()` will return null here? In that case we could probably return `e.getLocalizedMessage()`, and if either message is null, we should probably return the class name of the exception/cause.

Comment (shawnweeks) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AbstractRecordProcessor.java]: We're wrapping both exceptions directly above here so I think they're should always be a cause but it's possible the message might be null.

**Comment by shawnweeks** [comment]
So I'm breaking a test case in TestConvertRecord that checks to see if the flow file was modified when being routed to failure. Adding attributes creates a new flow file so while the contents haven't changed the object isn't the same.

**Comment by shawnweeks** [comment]
I can change the check to compare the contents of the flow file instead since that's really what matters.

**Comment by shawnweeks** [comment]
Rebased with master and moved WritesAttributes out of the abstract class for now.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestConvertRecord.java [hunk 2]
```diff
@@ -177,7 +176,7 @@ public void testWriteFailure() throws InitializationException {
         // Original FlowFile should be routed to 'failure' relationship without modification
         runner.assertAllFlowFilesTransferred(ConvertRecord.REL_FAILURE, 1);
         final MockFlowFile out = runner.getFlowFilesForRelationship(ConvertRecord.REL_FAILURE).get(0);
-        assertTrue(original == out);
+        out.assertContentEquals(original.toByteArray());
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestConvertRecord.java]: Same here, can you add an assert here to verify you got the attribute set as expected?

**Comment by shawnweeks** [comment]
My build environment has died so I don't know if this will compile or not.

**Comment by mattyb149** [comment]
+1 LGTM, ran contrib-check and tested on a live NiFi, verified the error attribute is added with the appropriate error message. Thanks for the improvement! Merging to master

**Pull request closed without merging**